### ISSUE TITLE
TodoList fails to build

### ISSUE THREAD

**Issue by tsukhu** [Microsoft/reactxp#906]
While the other samples compile and run , the TodoList fails to compile
```
 npm run start-web
.....
ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\app\AppBootstrapper.tsx
./src/ts/app/AppBootstrapper.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\app\AppBootstrapper.tsx(33,42)
      TS2345: Argument of type 'Element' is not assignable to parameter of type 'ReactElement<any>'.
  Types of property 'type' are incompatible.
    Type 'string | ComponentClass<any, any> | StatelessComponent<any>' is not assignable to type 'string | ComponentClass<any> | StatelessComponent<any>'.
      Type 'ComponentClass<any, any>' is not assignable to type 'string | ComponentClass<any> | StatelessComponent<any>'.
        Type 'ComponentClass<any, any>' is not assignable to type 'ComponentClass<any>'.
          Type 'Component<any, any, any>' is not assignable to type 'Component<any, ComponentState, any>'.
            Property 'context' is missing in type 'Component<any, any, any>'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\app\AppBootstrapper.tsx
./src/ts/app/AppBootstrapper.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\app\AppBootstrapper.tsx(59,13)
      TS2605: JSX element type 'RootView' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'RootView'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\controls\SimpleDialog.tsx
./src/ts/controls/SimpleDialog.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\controls\SimpleDialog.tsx(150,13)
      TS2605: JSX element type 'Modal' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'Modal'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\AccountMenuButton.tsx
./src/ts/views/AccountMenuButton.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\AccountMenuButton.tsx(110,21)
      TS2605: JSX element type 'SimpleMenu' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'SimpleMenu'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(90,21)
      TS2605: JSX element type 'NavigatorImpl' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'NavigatorImpl'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(101,21)
      TS2605: JSX element type 'TopBarComposite' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'TopBarComposite'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(152,17)
      TS2605: JSX element type 'TopBarStack' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'TopBarStack'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(166,21)
      TS2605: JSX element type 'TodoListPanel' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'TodoListPanel'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(180,24)
      TS2605: JSX element type 'ViewTodoPanel' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'ViewTodoPanel'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx
./src/ts/views/RootView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\RootView.tsx(203,20)
      TS2605: JSX element type 'TodoCompositeView' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'TodoCompositeView'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoCompositeView.tsx
./src/ts/views/TodoCompositeView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoCompositeView.tsx(52,21)
      TS2605: JSX element type 'TodoListPanel' is not a constructor function for JSX elements.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoCompositeView.tsx
./src/ts/views/TodoCompositeView.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoCompositeView.tsx(72,17)
      TS2605: JSX element type 'ViewTodoPanel' is not a constructor function for JSX elements.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoListPanel.tsx
./src/ts/views/TodoListPanel.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TodoListPanel.tsx(166,13)
      TS2605: JSX element type 'TodoListItem' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'TodoListItem'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TopBarComposite.tsx
./src/ts/views/TopBarComposite.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\TopBarComposite.tsx(100,21)
      TS2605: JSX element type 'AccountMenuButton' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'AccountMenuButton'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\ViewTodoPanel.tsx
./src/ts/views/ViewTodoPanel.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\ViewTodoPanel.tsx(76,13)
      TS2605: JSX element type 'SimpleDialog' is not a constructor function for JSX elements.
  Property 'context' is missing in type 'SimpleDialog'.

ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\ViewTodoPanel.tsx
./src/ts/views/ViewTodoPanel.tsx
[tsl] ERROR in C:\workspace\reactxp\samples\TodoList\src\ts\views\ViewTodoPanel.tsx(95,23)
      TS2345: Argument of type 'Element' is not assignable to parameter of type 'ReactElement<ViewProps>'.
  Types of property 'type' are incompatible.
    Type 'string | ComponentClass<any, any> | StatelessComponent<any>' is not assignable to type 'string | ComponentClass<ViewProps> | StatelessComponent<ViewProps>'.
      Type 'ComponentClass<any, any>' is not assignable to type 'string | ComponentClass<ViewProps> | StatelessComponent<ViewProps>'.
        Type 'ComponentClass<any, any>' is not assignable to type 'ComponentClass<ViewProps>'.
          Type 'Component<any, any, any>' is not assignable to type 'Component<ViewProps, ComponentState, any>'.
            Property 'context' is missing in type 'Component<any, any, any>'.
```

**Comment by erictraut** [comment]
I'm not able to repro the problem. Please make sure you are running a recent version of node/npm and that you've run ```npm install``` to download all of the dependencies for the TodoList app. If those suggestions don't solve the problem, we'll need to debug further.

**Comment by erictraut** [comment]
Any update?

**Comment by a-tarasyuk** [comment]
@tsukhu do you have `@types/react` in `package.json`? If yes, which version do you use?

**Comment by tsukhu** [comment]
@a-tarasyuk I am just using the master branch sources to build . The package.json for this sample has the following

  "@types/react": "^16.4.14",

**Comment by darrencruse** [comment]
just fwiw I'm new to ReactXP I pulled down the repo and tried the TodoList sample yesterday and the web version worked for me without problem (I had a little trouble with the android side but not web!).

I was on node v10.13.0.

(don't have my machine in front of me or I could try and offer more info...)

**Comment by amritbhandari** [comment]
I had cloned the repo couples of days ago and ran into building issue with the samples, but it got fixed when I moved to node 8.12.0.

For the TodoList sample the web and iOS versions worked fine without any issues, but I had issues with the Android version.

I finally did manage to get the Android version working, but only by making the following changes to it

- Gradle wrapper upgraded from: 3.3 to: 4.4
- Adding google() to repositories in the top level gradle file
- com.android.tools.build:gradle from: 2.2.3 to: 3.1.4
- compileSdkVersion from: 23 to: 27
- buildToolsVersion from: 25.0.0 to 27.0.3
- com.android.support:appcompat-v7 from: 23.0.1 to 27.1.1

Here is my system details
macOS: 10.13.6 (17G4015)
Android studio: 3.2.1
JDK: 1.8.0.191 oracle
Node: 8.12.0
Npm: 6.4.1

**Comment by deregtd** [comment]
@amritbhandari Awesome!  Can you possibly PR those changes back to us?

**Comment by P-Blackburn** [comment]
I downloaded the latest commit and attempted to build and deploy the TodoList sample following the instructions. in the [README.md](https://github.com/Microsoft/reactxp/blob/8e2e8d269c9b41a69b0eb12e89010eddd7433396/samples/TodoList/README.md).  

The only way I could get it to work in the end was to manually create an 'assets' folder e.g.
`mkdir .\TodoList\Android\app\src\main\assets`

AND then used the following yarn command in a shell on the TodoList folder:

`yarn react-native bundle --platform android --dev false --entry-file index.android.js --bundle-output android/app/src/main/assets/index.android.bundle --assets-dest android/app/src/main/res`

This appears to create a file 'index.android.bundle' in the assets folder - and then Android Studio is able to build deploy and debug it successfully to an attached Android device.

I do not yet fully understand what yarn was doing for me to get things to work - but somewhere I suspect that the existing npm scripts are currently failing to create the requisite files.

Can anyone shed any light on this?

**Comment by mikehardy** [comment]
I'm not sure what's going on in Windows, but on macOS 10.14.3, Android Studio 3.3 and a fresh checkout of reactxp master, the TodoList sample just ran fine for me per the instructions.

With mismatched Android Studio + React Native versions there was a bundler issue recently for Android builds that required you to pre-bundle (which is what you did with that `react-native bundle` command - you created your distribution Javascript bundle manually in a specific location, as opposed to using the one the react-native metro packager would otherwise make). That pre-bundle requirement was fixed somewhere in the react-native 0.57.3 time frame though I think? And it only applied after you moved to modern Gradle + Android Studio combinations (like Android Studio 3.2.+ with Gradle 4.7+ I think). Either way it doesn't apply anymore with current Android toolchains, on Mac and Linux at least.

Not sure if this will help or not for the Windows side, but if it can serve to at least let you know it should be working, and does work as-is on other platforms, maybe that's helpful

**Comment by P-Blackburn** [comment]
Thank you for the confirmation that it _should_ be working and is indeed working on macOS - I am indeed on Windows10 x64 - and will test later on Debian 9 x64...

I (re)-started from a brand new (clean) installation of Windows 10 x64 - and took baby steps :) as follows
I have installed: 

1. VSCode 1.30.2
2. Android Studio 3.3
3. NodeJs 10.15.1
4. npm 6.4.1

Cloneed the reactxp repository to 5a44ec8 (today)

**Web Build--**
Trying first to build the TodoList project for the web executing `npm run start-web` in the TodoList folder - results in a bunch of warnings  about dependencies....

**Are these warnings critical\important?** and is the correct thing to do to resolve them to faff about in the package.json file to fix dependencies on specific versions or is that step not needed?  -  I am assuming that it is not needed as the web version of TodoList appears to kind of work.

**What did not work** was the ability to delete a created task - I am assuming that this is unfinished coding?

**Android Build--**
Similarly with `npm run start-android` there are a bunch of npm warnings about required versions not being installed etc. 

In a second terminal I start `npm run start`

And then in Android Studio I try to build - Android Studio reports errors and the `npm run start` falls over with the attached log
[2019-01-31T00_37_33_893Z-debug.log](https://github.com/Microsoft/reactxp/files/2815201/2019-01-31T00_37_33_893Z-debug.log)
